File: src/lib/pitono/src/base_given.py
from typing import List, Dict, Any, Callable, Optional
from .base_setup import BaseSetup
from .base_when import BaseWhen
from .base_then import BaseThen
from .pitono_types import ITestResourceConfiguration, ITestArtifactory

class BaseGiven(BaseSetup):
    def __init__(
        self,
        key: str,
        features: List[str],
        whens: List[BaseWhen],
        thens: List[BaseThen],
        given_cb: Callable,
        initial_values: Any
    ):
        # Map whens to actions and thens to checks
        super().__init__(
            features,
            whens,
            thens,
            given_cb,
            initial_values
        )
        self.key = key
    
    async def given_that(
        self,
        subject: Any,
        test_resource_configuration: ITestResourceConfiguration,
        artifactory: ITestArtifactory,
        given_cb: Callable,
        initial_values: Any
    ) -> Any:
        # Default implementation that can be overridden
        # This should call setup_that from BaseSetup
        return await self.setup_that(
            subject,
            test_resource_configuration,
            artifactory,
            given_cb,
            initial_values
        )
    
    async def setup_that(
        self,
        subject: Any,
        test_resource_configuration: ITestResourceConfiguration,
        artifactory: ITestArtifactory,
        setup_cb: Callable,
        initial_values: Any
    ) -> Any:
        # Default implementation
        if callable(setup_cb):
            return setup_cb(subject, initial_values)
        return subject
    
    # Alias setup to give for BDD pattern
    async def give(
        self,
        subject: Any,
        key: str,
        test_resource_configuration: ITestResourceConfiguration,
        tester: Callable[[Any], bool],
        artifactory: Optional[ITestArtifactory] = None,
        suite_ndx: Optional[int] = None
    ) -> Any:
        return await super().setup(
            subject,
            key,
            test_resource_configuration,
            tester,
            artifactory,
            suite_ndx
        )

File: src/lib/kafe/src/main/java/kafe/BaseGiven.java
package kafe;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

/**
 * BaseGiven extends BaseSetup to support BDD pattern.
 * It reuses all Setup functionality but with BDD naming.
 */
public abstract class BaseGiven<I, S, R> extends BaseSetup<I, S, R> {
    public BaseGiven(
        List<String> features,
        List<BaseWhen<S, R>> whens,  // These will be processed as actions
        List<BaseThen<R>> thens,     // These will be processed as checks
        Function<S, R> givenCb,
        Object initialValues
    ) {
        // Convert whens to actions and thens to checks
        super(
            features,
            new ArrayList<>(whens),
            new ArrayList<>(thens),
            givenCb,
            initialValues
        );
    }
    
    // Alias setupThat to givenThat for BDD pattern
    @Override
    public abstract R setupThat(
        S subject,
        ITTestResourceConfiguration testResourceConfiguration,
        Function<String, Object> artifactory,
        Function<S, R> setupCB,
        Object initialValues
    );
    
    // Alias setup to give for BDD pattern
    public R give(
        S subject,
        String key,
        ITTestResourceConfiguration testResourceConfiguration,
        Function<Object, Boolean> tester,
        Function<String, Object> artifactory,
        int suiteNdx
    ) throws Exception {
        return super.setup(subject, key, testResourceConfiguration, tester, artifactory, suiteNdx);
    }
}
